### PR TITLE
Feature: Staking

### PR DESCRIPTION
**Pull request by talekhinezh**

### CODE REVIEW

**Review by dhedey — APPROVED** [review]
Looks good!

I have quite a big worry around Map/HashMap ordering in the Java side. I suggest that we use a `LinkedHashMap` on the Java side. I'm wary of using a `TreeMap` because the ordering of the types between Rust and Java may not be obvious, and may potentially be different if they define different ordering semantics. Instead, we can make `Map` in Java SBOR use a `LinkedHashMap` behind the scenes, and just preserve the order from rust.

RE PR co-ordination - would appreciate if you co-ordinate with Alexandru RE ordering. I think he's planning to merge first tomorrow. His PR actually uses a further version of develop than yours, so you will need to merge in a later version of develop to merge this PR.

File: core-rust/core-api-server/Cargo.toml [hunk 1]
```diff
@@ -23,13 +23,13 @@ parking_lot = { version = "0.12" }
...
 # Ensure this version is also identically updated in ../state-manager/Cargo.toml
 #
-sbor = { git = "https://github.com/radixdlt/radixdlt-scrypto", tag = "develop-dac211", features = ["serde", "indexmap"] }
-transaction = { git = "https://github.com/radixdlt/radixdlt-scrypto", tag = "develop-dac211" }
-radix-engine-interface = { git = "https://github.com/radixdlt/radixdlt-scrypto", tag = "develop-dac211", features = ["serde"]  }
-radix-engine = { git = "https://github.com/radixdlt/radixdlt-scrypto", tag = "develop-dac211" }
-radix-engine-constants = { git = "https://github.com/radixdlt/radixdlt-scrypto", tag = "develop-dac211" }
-radix-engine-stores = { git = "https://github.com/radixdlt/radixdlt-scrypto", tag = "develop-dac211" }
-utils = { git = "https://github.com/radixdlt/radixdlt-scrypto", tag = "develop-dac211" }
+sbor = { git = "https://github.com/radixdlt/radixdlt-scrypto", tag = "staking-5b6e06", features = ["serde", "indexmap"] }
```
[the remark this answers is not in the document]
Comment (dhedey) [reply core-rust/core-api-server/Cargo.toml:26]: Alexandru's PR merges a more recent version of develop, so I suggest potentially merging after him tomorrow :)

File: core-rust-bridge/src/main/java/com/radixdlt/statecomputer/commit/NextEpoch.java [hunk 2]
```diff
@@ -65,12 +65,13 @@
...
 import com.google.common.reflect.TypeToken;
+import com.radixdlt.rev2.SystemAddress;
 import com.radixdlt.sbor.codec.CodecMap;
 import com.radixdlt.sbor.codec.StructCodec;
 import com.radixdlt.utils.UInt64;
-import java.util.Set;
+import java.util.Map;
 
-public record NextEpoch(Set<Validator> validators, UInt64 epoch) {
+public record NextEpoch(Map<SystemAddress, Validator> validators, UInt64 epoch) {
```
[the remark this answers is not in the document]
Comment (talekhinezh) [reply core-rust-bridge/src/main/java/com/radixdlt/statecomputer/commit/NextEpoch.java:74]: How about `ImmutableMap` which maintains order.

File: core-rust-bridge/src/main/java/com/radixdlt/statecomputer/commit/Validator.java [hunk 3]
```diff
@@ -66,11 +66,11 @@
 
 import com.google.common.reflect.TypeToken;
 import com.radixdlt.crypto.ECDSASecp256k1PublicKey;
-import com.radixdlt.rev2.SystemAddress;
+import com.radixdlt.rev2.Decimal;
 import com.radixdlt.sbor.codec.CodecMap;
 import com.radixdlt.sbor.codec.StructCodec;
 
-public record Validator(SystemAddress address, ECDSASecp256k1PublicKey key) {
+public record Validator(ECDSASecp256k1PublicKey key, Decimal stake) {
```
[the remark this answers is not in the document]
Comment (talekhinezh) [reply core-rust-bridge/src/main/java/com/radixdlt/statecomputer/commit/Validator.java:73]: Sounds good.

File: core-rust/core-api-server/Cargo.toml (hunk 1 above)
[the remark this answers is not in the document]
Comment (talekhinezh) [reply core-rust/core-api-server/Cargo.toml:26]: It was merged but is merged a bit too far in the future. Can definitely merge after Alexandru, which may or may not be a develop tag.

File: core-rust/core-api-server/core-api-schema.yaml [hunk 4]
```diff
@@ -2017,16 +2017,29 @@ components:
 #######################################
 # GENERAL / SHARED MODELS - validators #
 #######################################
-    Validator:
+    ValidatorEntry:
```
[the remark this answers is not in the document]
Comment (talekhinezh) [reply core-rust/core-api-server/core-api-schema.yaml:2020]: Yup, definitely possible.

File: core-rust/core-api-server/core-api-schema.yaml [hunk 5]
```diff
@@ -2017,16 +2017,29 @@ components:
...
+      required:
+        - key
+        - stake
+      properties:
         key:
           $ref: '#/components/schemas/EcdsaSecp256k1PublicKey'
+        stake:
+          type: string
+          description: |
+            The string-encoded decimal representing the amount of XRD staked.
```
[the remark this answers is not in the document]
Comment (talekhinezh) [reply core-rust/core-api-server/core-api-schema.yaml:2041]: Sounds Good.

File: core-rust-bridge/src/main/java/com/radixdlt/statecomputer/commit/NextEpoch.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (dhedey) [reply core-rust-bridge/src/main/java/com/radixdlt/statecomputer/commit/NextEpoch.java:74]: Yes, sounds perfect: https://stackoverflow.com/questions/3810738/google-collections-immutablemap-iteration-order

**Comment by Theoklitos** [comment]
Regarding the failing `ephemeral deploy and test` action: We can ignore it. I will ask for a radix-engine-toolkit update, but this should not block this PR

**Pull request merged**